docs/cudf/source/developer_guide/developer_guide.md

```diff
@@ -0,0 +1,129 @@
+Under the hood, all of cuDF's functionality relies on the CUDA-accelerated `libcudf` C++ library.
+Thus, cuDF's internals are designed to efficiently and robustly map pandas APIs to `libcudf` functions.
+For more information about the `libcudf` library, a good starting point is the
+[developer guide](https://github.com/rapidsai/cudf/blob/main/cpp/docs/DEVELOPER_GUIDE.md).
+
+This document assumes familiarity with the
+[overall contributing guide](https://github.com/rapidsai/cudf/blob/main/CONTRIBUTING.md).
+The goal of this document is to provide more specific guidance for Python developers.
+It covers the structure of the Python code and discusses best practices.
+Additionally, it includes longer sections on more specific topics like testing and benchmarking.
+More specific information on these can be found in the pages below:
+
+```{toctree}
+:maxdepth: 1
+
+library_design
+```
+
+The rest of this document focuses on a higher-level overview of best practices in cuDF.
+
+## Directory structure and file naming
+
+cuDF generally presents the same importable modules and subpackages as pandas.
+All Cython code is contained in `python/cudf/cudf/_lib`.
+
+**Open question**: Should we start enforcing a stricter file/directory layout? Any suggestions?
+
+
+## Code style
+
+cuDF employs a number of linters to ensure consistent style across the code base:
+
+- [`flake8`](https://github.com/pycqa/flake8) checks for general code formatting compliance. 
+- [`black`](https://github.com/psf/black) is an automatic code formatter.
+- [`isort`](https://pycqa.github.io/isort/) ensures imports are sorted consistently.
+- [`mypy`](http://mypy-lang.org/) performs static type checking.
+  In conjunction with [type hints](https://docs.python.org/3/library/typing.html),
+  `mypy` can help catch various bugs that are otherwise difficult to find.
+- [`pydocstyle`](https://github.com/PyCQA/pydocstyle/) lints docstring style.
+
+Configuration information for these tools is contained in a number of files.
+The primary source of truth is the `.pre-commit-config.yaml` file at the root of the repo.
+As described in the
+[overall contributing guide](https://github.com/rapidsai/cudf/blob/main/CONTRIBUTING.md),
+we recommend using [`pre-commit`](https://pre-commit.com/) to manage all linters.
+
+## Deprecating and removing code
+
+cuDF generally follows the policy of deprecating code for one release prior to removal.
+For example, if we decide to remove an API during the 22.08 release cycle,
+it will be marked as deprecated in the 22.08 release and removed in the 22.10 release.
+All internal usage of deprecated APIs in cuDF should be removed when the API is deprecated.
+This prevents users from encountering unexpected deprecation warnings when using other (non-deprecated) APIs.
+The documentation for the API should also be updated to reflect its deprecation.
+
+When deprecating an API, developers should open a corresponding GitHub issue to track the API removal.
```

Comment:
Hmm that's an interesting question that I haven't considered. Do you think it's worth having a checklist with the items listed here (make sure all calling code is updated, remove tests, etc)?

Reply:
I had a look to see how often APIs are deprecated, which is with some regularity. At the moment it doesn't look like there is consistency on whether a tracking issue is created for their subsequent removal. I suspect because that requirement was not previously clearly defined.

While there's a `deprecation` label, it is not consistently applied (there are 24 PRs tagged with deprecation but only 8 issues).

I think an issue template would help with remembering to fill in the issue. I suppose one could even have an automation that enforces that deprecation PRs are linked to by a deprecation issue before merge (but that needs more scripting...).

Maybe something simple like:

```
---
name: Deprecated API tracker
about: Create a tracking issue for removal of a deprecated API
title: "[DEP]"
labels: "deprecation"
assignees: ''

---

** Tracking issue for removal of `FILL_IN_API_CALL`, deprecated in `PR-NUMBER`

- [ ] Calling code removed
- [ ] Tests removed
- [ ] Documentation updated

```

Reply:
I am both the person who originally suggested this requirement, and the most frequent violator of it. In hindsight, I am wondering whether this requirement is actually useful. In practice, I have instead been operating using a different strategy. Since our deprecation cycle is a single release, as part of the release cycle I basically do a post-portem search for `FutureWarning` and delete. Do you think that is sufficient, as opposed to the overhead of creating many issues?

CC @shwina @bdice who were involved in the original discussions when I proposed using a label for deprecations.

Reply:
> Since our deprecation cycle is a single release, as part of the release cycle I basically do a post-portem search for `FutureWarning` and delete. Do you think that is sufficient, as opposed to the overhead of creating many issues?

I agree the requirement to create an issue is unnecessary and too much overhead/churn. We sometimes let deprecations slip for 2 releases if we don't have time to rip them out, and that's okay too. We have been in a good habit of keeping things moving from deprecation to removal and I wouldn't over-codify this at present.

Reply:
Agreed with @bdice